mouse.lisp

```diff
@@ -0,0 +1,124 @@
+        (xlib:global-pointer-position *display*)
+      (let* ((leniency-offset 2)
+             (x (xlib:drawable-x (window-parent window)))
+             (w (xlib:drawable-width (window-parent window)))
+             (min-x (- x *float-window-border* leniency-offset))
+             (max-x (+ x w *float-window-border* leniency-offset))
+             (y (xlib:drawable-y (window-parent window)))
+             (h (xlib:drawable-height (window-parent window)))
+             (min-y (- y *float-window-title-height* leniency-offset))
+             (max-y (+ y h *float-window-border* leniency-offset)))
+        (and (<= min-x mouse-x max-x)
+             (<= min-y mouse-y max-y)))))
+  (:method ((window tile-window))
+    (let ((frame (window-frame window)))
+      (mouse-inside-frame-p frame))))
+
+(defun mouse-banish-window (window)
+  "Move mouse pointer to edge of a window."
+  (let* ((min-x (xlib:drawable-x (window-parent window)))
+         (max-x (+ min-x (xlib:drawable-width (window-parent window))))
+         (new-x (if (minusp *mouse-follow-banish-x-offset*)
+                    (+ max-x *mouse-follow-banish-x-offset*)
+                    (+ min-x *mouse-follow-banish-x-offset*)))
+         (min-y (xlib:drawable-y (window-parent window)))
+         (max-y (+ min-y (xlib:drawable-height (window-parent window))))
+         (new-y (if (minusp *mouse-follow-banish-y-offset*)
+                    (+ max-y *mouse-follow-banish-y-offset*)
+                    (+ min-y *mouse-follow-banish-y-offset*))))
+    (ratwarp (clamp new-x min-x max-x)
+             (clamp new-y min-y max-y))))
+
+(defun mouse-handle-focus-frame (current-frame last-frame)
+  "Move mouse when moving frames."
+  (when (eq *mouse-follow-policy* :follow)
+    (unless (or (eq current-frame last-frame)
+                (mouse-inside-frame-p current-frame))
+      (let ((current-window (frame-window current-frame)))
+        (if current-window
+            (mouse-banish-window current-window)
+            (mouse-banish-frame current-frame))))))
+
+(defun mouse-handle-split-frame (first-frame)
+  "Reposition the mouse when a frame is created."
+  (when (eq *mouse-follow-policy* :follow)
+    (mouse-banish-frame first-frame)))
+
+(defun mouse-handle-remove-split (current-frame)
+  "Reposition the mouse when a frame is removed."
+  (when (eq *mouse-follow-policy* :follow)
+    (mouse-banish-frame current-frame)))
+
+(defun mouse-handle-focus-window (current-window last-window)
+  "Move mouse for floating windows."
+  (when (eq *mouse-follow-policy* :follow)
+    (unless (or (eq current-window last-window)
+                (mouse-inside-window-p current-window))
+      (mouse-banish-window current-window))))
+
+(defun mouse-handle-focus-group (group)
+  "Disable sloppy pointer when switching groups to prevent floating windows from
+getting stuck and banish to last window or frame."
+  (when (eq *mouse-follow-policy* :follow)
+    (let ((window (current-window)))
+      (if window
```

Comment:
:thought_balloon: You can combine the (let () (if ....)) with alexandria's if-let, which is already exported in the StumpWM package.

```lisp
(if-let ((window (current-window)))
  (mouse-banish-window window)
  (mouse-banish-frame (tile-group-current-frame group)))
```